File: backend/routes/inventory.py
# Inventory Management Routes
from flask import Blueprint, request, jsonify
from services.supabase_client import get_supabase
from datetime import datetime

bp = Blueprint('inventory', __name__)
supabase = get_supabase()

@bp.route('', methods=['GET'])
def get_inventory():
    """Get all inventory items"""
    try:
        # Query parameters
        fabric_type = request.args.get('fabric_type', '')
        low_stock = request.args.get('low_stock', 'false')
        
        query = supabase.table('inventory').select('*')
        
        if fabric_type:
            query = query.eq('fabric_type', fabric_type)
        
        response = query.order('last_updated', desc=True).execute()
        
        # Filter low stock items if requested
        data = response.data
        if low_stock.lower() == 'true':
            data = [item for item in data if float(item['quantity_meters']) < 50]
        
        return jsonify({
            'success': True,
            'data': data,
            'count': len(data)
        }), 200
        
    except Exception as e:
        return jsonify({'success': False, 'error': str(e)}), 500

@bp.route('/<inventory_id>', methods=['GET'])
def get_inventory_item(inventory_id):
    """Get single inventory item"""
    try:
        response = supabase.table('inventory').select('*').eq('id', inventory_id).execute()
        
        if not response.data:
            return jsonify({'success': False, 'error': 'Inventory item not found'}), 404
        
        return jsonify({
            'success': True,
            'data': response.data[0]
        }), 200
        
    except Exception as e:
        return jsonify({'success': False, 'error': str(e)}), 500

@bp.route('', methods=['POST'])
def add_inventory():
    """Add new inventory item"""
    try:
        data = request.get_json()
        
        # Validate required fields
        required_fields = ['fabric_type', 'fabric_color', 'quantity_meters', 'rate_per_meter']
        for field in required_fields:
            if field not in data:
                return jsonify({'success': False, 'error': f'{field} is required'}), 400
        
        # Check if fabric already exists
        existing = supabase.table('inventory').select('*').eq('fabric_type', data['fabric_type']).eq('fabric_color', data['fabric_color']).execute()
        
        if existing.data:
            # Update existing quantity
            inventory_id = existing.data[0]['id']
            current_quantity = float(existing.data[0]['quantity_meters'])
            new_quantity = current_quantity + float(data['quantity_meters'])
            
            response = supabase.table('inventory').update({
                'quantity_meters': new_quantity,
                'rate_per_meter': data['rate_per_meter'],
                'last_updated': datetime.now().isoformat()
            }).eq('id', inventory_id).execute()
            
            return jsonify({
                'success': True,
                'message': 'Inventory updated (quantity added)',
                'data': response.data[0]
            }), 200
        else:
            # Create new inventory item
            inventory_data = {
                'fabric_type': data['fabric_type'],
                'fabric_color': data['fabric_color'],
                'quantity_meters': data['quantity_meters'],
                'rate_per_meter': data['rate_per_meter'],
                'location': data.get('location', '')
            }
            
            response = supabase.table('inventory').insert(inventory_data).execute()
            
            return jsonify({
                'success': True,
                'message': 'Inventory item created successfully',
                'data': response.data[0]
            }), 201
        
    except Exception as e:
        return jsonify({'success': False, 'error': str(e)}), 500

@bp.route('/<inventory_id>', methods=['PUT'])
def update_inventory(inventory_id):
    """Update inventory item"""
    try:
        data = request.get_json()
        
        # Check if item exists
        existing = supabase.table('inventory').select('*').eq('id', inventory_id).execute()
        if not existing.data:
            return jsonify({'success': False, 'error': 'Inventory item not found'}), 404
        
        # Update inventory
        update_data = {'last_updated': datetime.now().isoformat()}
        
        if 'quantity_meters' in data:
            update_data['quantity_meters'] = data['quantity_meters']
        if 'rate_per_meter' in data:
            update_data['rate_per_meter'] = data['rate_per_meter']
        if 'location' in data:
            update_data['location'] = data['location']
        if 'fabric_color' in data:
            update_data['fabric_color'] = data['fabric_color']
        
        response = supabase.table('inventory').update(update_data).eq('id', inventory_id).execute()
        
        return jsonify({
            'success': True,
            'message': 'Inventory updated successfully',
            'data': response.data[0]
        }), 200
        
    except Exception as e:
        return jsonify({'success': False, 'error': str(e)}), 500

@bp.route('/<inventory_id>', methods=['DELETE'])
def delete_inventory(inventory_id):
    """Delete inventory item"""
    try:
        # Check if item exists
        existing = supabase.table('inventory').select('*').eq('id', inventory_id).execute()
        if not existing.data:
            return jsonify({'success': False, 'error': 'Inventory item not found'}), 404
        
        # Delete inventory
        supabase.table('inventory').delete().eq('id', inventory_id).execute()
        
        return jsonify({
            'success': True,
            'message': 'Inventory item deleted successfully'
        }), 200
        
    except Exception as e:
        return jsonify({'success': False, 'error': str(e)}), 500

@bp.route('/low-stock', methods=['GET'])
def get_low_stock():
    """Get low stock items (< 50 meters)"""
    try:
        response = supabase.table('inventory').select('*').execute()
        
        low_stock_items = [item for item in response.data if float(item['quantity_meters']) < 50]
        
        return jsonify({
            'success': True,
            'data': low_stock_items,
            'count': len(low_stock_items)
        }), 200
        
    except Exception as e:
        return jsonify({'success': False, 'error': str(e)}), 500

@bp.route('/summary', methods=['GET'])
def get_inventory_summary():
    """Get inventory summary statistics"""
    try:
        response = supabase.table('inventory').select('*').execute()
        
        total_items = len(response.data)
        low_stock_count = len([item for item in response.data if float(item['quantity_meters']) < 50])
        total_meters = sum(float(item['quantity_meters']) for item in response.data)
        total_value = sum(float(item['quantity_meters']) * float(item['rate_per_meter']) for item in response.data)
        
        return jsonify({
            'success': True,
            'data': {
                'total_items': total_items,
                'low_stock_count': low_stock_count,
                'total_meters': round(total_meters, 2),
                'total_value': round(total_value, 2)
            }
        }), 200
        
    except Exception as e:
        return jsonify({'success': False, 'error': str(e)}), 500

File: backend-node/routes/inventory.js
// Inventory Routes
const express = require('express');
const router = express.Router();
const store = require('../data/store');

// GET /api/inventory - List all inventory items
router.get('/', (req, res) => {
    try {
        let inventory = store.findAll('inventory');

        // Filter by fabric_type
        if (req.query.fabric_type) {
            inventory = inventory.filter(i => i.fabric_type === req.query.fabric_type);
        }

        // Filter low stock (quantity < 50)
        if (req.query.low_stock === 'true') {
            inventory = inventory.filter(i => parseFloat(i.quantity_meters) < 50);
        }

        res.json({ success: true, data: inventory });
    } catch (error) {
        res.status(500).json({ success: false, error: error.message });
    }
});

// GET /api/inventory/low-stock - Get low stock items
router.get('/low-stock', (req, res) => {
    try {
        const lowStock = store.findAll('inventory').filter(i => parseFloat(i.quantity_meters) < 50);
        res.json({ success: true, data: lowStock });
    } catch (error) {
        res.status(500).json({ success: false, error: error.message });
    }
});

// GET /api/inventory/summary - Get inventory summary
router.get('/summary', (req, res) => {
    try {
        const inventory = store.findAll('inventory');
        const totalItems = inventory.length;
        const totalMeters = inventory.reduce((sum, i) => sum + parseFloat(i.quantity_meters), 0);
        const totalValue = inventory.reduce(
            (sum, i) => sum + parseFloat(i.quantity_meters) * parseFloat(i.rate_per_meter), 0
        );

        res.json({
            success: true,
            data: {
                total_items: totalItems,
                total_meters: totalMeters,
                total_value: totalValue,
            },
        });
    } catch (error) {
        res.status(500).json({ success: false, error: error.message });
    }
});

// GET /api/inventory/:id - Get item by ID
router.get('/:id', (req, res) => {
    try {
        const item = store.findById('inventory', req.params.id);
        if (!item) {
            return res.status(404).json({ success: false, error: 'Inventory item not found' });
        }
        res.json({ success: true, data: item });
    } catch (error) {
        res.status(500).json({ success: false, error: error.message });
    }
});

// POST /api/inventory - Add inventory item
router.post('/', (req, res) => {
    try {
        const { fabric_type, fabric_color, quantity_meters, rate_per_meter, location } = req.body;

        if (!fabric_type || !fabric_color || !quantity_meters || !rate_per_meter) {
            return res.status(400).json({
                success: false,
                error: 'fabric_type, fabric_color, quantity_meters, and rate_per_meter are required',
            });
        }

        const item = store.create('inventory', {
            fabric_type,
            fabric_color,
            quantity_meters: parseFloat(quantity_meters),
            rate_per_meter: parseFloat(rate_per_meter),
            location: location || '',
        });

        res.status(201).json({ success: true, data: item });
    } catch (error) {
        res.status(500).json({ success: false, error: error.message });
    }
});

// PUT /api/inventory/:id - Update inventory item
router.put('/:id', (req, res) => {
    try {
        const updates = { ...req.body };
        if (updates.quantity_meters) updates.quantity_meters = parseFloat(updates.quantity_meters);
        if (updates.rate_per_meter) updates.rate_per_meter = parseFloat(updates.rate_per_meter);

        const updated = store.update('inventory', req.params.id, updates);
        if (!updated) {
            return res.status(404).json({ success: false, error: 'Inventory item not found' });
        }
        res.json({ success: true, data: updated });
    } catch (error) {
        res.status(500).json({ success: false, error: error.message });
    }
});

// DELETE /api/inventory/:id - Delete inventory item
router.delete('/:id', (req, res) => {
    try {
        const deleted = store.remove('inventory', req.params.id);
        if (!deleted) {
            return res.status(404).json({ success: false, error: 'Inventory item not found' });
        }
        res.json({ success: true, message: 'Inventory item deleted' });
    } catch (error) {
        res.status(500).json({ success: false, error: error.message });
    }
});

module.exports = router;
